### PR TITLE
feat(rslint_parser): simple jsx parser

### PR DESCRIPTION
**Pull request by xunilrj**
This PR implements:
    - Open/Close elements
    - Self closing elements

We still do not parse attributes, children etc...

# Formatter

Formatting was ignored, mainly because there are still a lot of parser errors.

# CheckpointedParser

```CheckpointedParser``` is a small improvement when we do tentative parsing. We have a rule to never abandon a marker if the parser is already advanced. This is not a problem when we checkpoint/reward. What this struct is doing is signalling that the function does abandon markers and should only be called under a checkpoint.

We will improve tentative parsing in future PRs, maybe helping the caller to deal with ambiguity somehow.

# Errors

Before generating diagnostics, we need to fix a problem with checkpoints, and reverting diagnostics on ```rewind```.

### CODE REVIEW

File: crates/rome_formatter/src/js/any/mod.rs [hunk 1]
```diff
@@ -43,3 +43,5 @@ mod root;
 mod statement;
 mod switch_clause;
 mod template_element;
+mod x_any_element;
```
Comment (MichaReiser) [inline crates/rome_formatter/src/js/any/mod.rs:46]: We need to change the codegen to support the `jsx` prefix.

File: crates/rslint_parser/Cargo.toml [hunk 2]
```diff
@@ -10,7 +10,9 @@ repository = "https://github.com/rslint/rslint"
 [dependencies]
 rslint_errors = { path = "../rslint_errors", version = "0.2.0" }
 rslint_syntax = { path = "../rslint_syntax", version = "0.1" }
-rslint_lexer = { path = "../rslint_lexer", version = "0.2", features = ["highlight"] }
+rslint_lexer = { path = "../rslint_lexer", version = "0.2", features = [
+    "highlight",
```
Comment (MichaReiser) [inline crates/rslint_parser/Cargo.toml:14]: Nit> Can we remove the `highlight` feature. The parser doesn't use it.

File: crates/rslint_parser/Cargo.toml [hunk 3]
```diff
@@ -19,6 +21,7 @@ indexmap = "1.8.0"
 cfg-if = "1.0.0"
 thiserror = "1.0.30"
 smallvec = { version = "1.8.0", features = ["union", "const_new"] }
+trace_caller = "0.2.0"
```
Comment (MichaReiser) [inline crates/rslint_parser/Cargo.toml:24]: For what is this dependency used?

File: crates/rslint_parser/src/parser.rs [hunk 4]
```diff
@@ -215,13 +215,12 @@ impl<'t> Parser<'t> {
...
     /// Consume the next token if `kind` matches.
+    // #[trace_caller::trace]
     pub fn bump(&mut self, kind: JsSyntaxKind) {
-        assert!(
-            self.eat(kind),
-            "expected {:?} but at {:?}",
-            kind,
-            self.cur()
-        );
+        let r = self.eat(kind);
```
Comment (MichaReiser) [inline crates/rslint_parser/src/parser.rs:220]: Nit: `eaten` or `is_present` instead of `r`

File: crates/rslint_parser/src/syntax/expr.rs [hunk 5]
```diff
@@ -1128,6 +1129,7 @@ fn parse_primary_expression(p: &mut Parser, context: ExpressionContext) -> Parse
         T!['('] => parse_parenthesized_expression(p, context).unwrap(),
         T!['['] => parse_array_expr(p).unwrap(),
         T!['{'] if context.is_object_expression_allowed() => parse_object_expression(p).unwrap(),
+        T![<] if context.is_object_expression_allowed() => try_parse_jsx_expression(p).unwrap(),
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1132]: Why is the test for `is_object_expression_allowed` needed. This, for example, is valid JSX

```
<test></test>
```

But this isn't the case for objects:

```
{test}
```
because this should parse as a block statement

File: crates/rslint_parser/src/syntax/expr.rs [hunk 6]
```diff
@@ -1562,6 +1564,7 @@ fn parse_call_expression_rest(
         // (() => { a }).a<A, B, C>()
         // (() => a)<A, B, C>();
         if TypeScript.is_supported(p) && p.at(T![<]) {
+            println!("HERE");
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1567]: Delete

File: crates/rslint_parser/src/syntax/expr.rs (hunk 5 above)
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/expr.rs:1132]: We should probably card this by a check if `Jsx` is enabled (at least, if TypeScript` is enabled

File: crates/rslint_parser/src/syntax/expr.rs [hunk 7]
```diff
@@ -1692,12 +1695,23 @@ pub(super) fn parse_unary_expr(p: &mut Parser, context: ExpressionContext) -> Pa
...
-        return TypeScript.parse_exclusive_syntax(
-            p,
-            |p| parse_ts_type_assertion_expression(p, context),
-            |p, assertion| ts_only_syntax_error(p, "type assertion", assertion.range(p).as_range()),
-        );
+        let jsx = if context.is_object_expression_allowed() {
+            try_parse_jsx_expression(p)
+        } else {
+            Absent
+        };
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:1704]: We should test if `Jsx` is enabled or not rather than testing for `is_object_expression_allowed`

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        return ParsedSyntax::Absent;
+    }
+
+    let mut p = CheckpointedParser::new(p);
+    let syntax = parse_jsx_expression(&mut p);
+    let (p, checkpoint) = p.unwrap();
+
+    if syntax.is_absent() {
+        p.rewind(checkpoint);
+    }
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:39]: How about if `CheckpointedParser` offers a `try_parse` method that automatically rewinds if `syntax` is absent? It would avoid exposing the internals 

Ideally, `try_parse` could even be implemented on the `Parser` itself and internally create the `CheckpointedParser`.

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+// }
+
+// test jsx jsx_element_on_arrow_function
+// const f = () => <div></div>;
+// const f = () => (<div></div>);
+
+// test jsx jsx_element_as_statements
+// <div />
+fn parse_jsx_expression(p: &mut CheckpointedParser<'_, '_>) -> ParsedSyntax {
+    let m = p.parser.start();
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:56]: Would it make sense to implement `Deref` on `CheckpointedParser`?

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+// test jsx jsx_element_open_close
+// function f() {
+//     return <div></div>
+// }
+
+// test jsx jsx_element_self_close
+// function f() {
+//     return <div />
+// }
+fn parse_jsx_element(p: &mut CheckpointedParser<'_, '_>) -> ParsedSyntax {
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:75]: I'm not sure if the `CheckpointedParser` will work out for you because you probably want to reuse the logic when parsing the children that can contain `JsxElement`. The parser won't need to guess inside of a `JsxElement`

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+// const f = () => <div></div>;
+// const f = () => (<div></div>);
+
+// test jsx jsx_element_as_statements
+// <div />
+fn parse_jsx_expression(p: &mut CheckpointedParser<'_, '_>) -> ParsedSyntax {
+    let m = p.parser.start();
+
+    if parse_jsx_element(p).is_absent() {
+        m.abandon(p.parser);
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:59]: You can use 


```rs
parse_jsx_element(p).map(|element| {
  let m =element.precede();
  Present(m.complete(p, JSX_ELEMENT_EXPRESSION))
});
```

Instead of having to create a marker and then abandoning it.

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+//     return <div />
+// }
+fn parse_jsx_element(p: &mut CheckpointedParser<'_, '_>) -> ParsedSyntax {
+    let m = p.parser.start();
+    match parse_jsx_element_head(p, m) {
+        ParsedSyntax::Present(opening_marker)
+            if opening_marker.kind() == JsSyntaxKind::JSX_OPENING_ELEMENT =>
+        {
+            let element_marker = opening_marker.precede(p.parser);
+            let closing_marker = parse_jsx_closing_element(p);
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:82]: Doesn't the closing element need the name to be able to match if the closing element matches the opening element?

File: xtask/codegen/js.ungram [hunk 13]
```diff
@@ -2136,3 +2139,52 @@ TsTypeArguments =
...
+	JsxElement 
+	| JsxSelfClosingElement
+
+
+// <a />
+JsxSelfClosingElement = 
+	'<'
+	name: JsxAnyElementName
+	'/'
+	'>'
```
Comment (MichaReiser) [inline xtask/codegen/js.ungram:2162]: We should bump this as a single token.

File: xtask/codegen/js.ungram [hunk 14]
```diff
@@ -2136,3 +2139,52 @@ TsTypeArguments =
...
+	closing_element: JsxClosingElement
+
+JsxOpeningElement = 
+	'<'
+	name: JsxAnyElementName	
+	'>'
+	
+JsxClosingElement = 
+	'<'
+	'/'
```
Comment (MichaReiser) [inline xtask/codegen/js.ungram:2176]: Use a single token

File: crates/rome_formatter/src/js/any/mod.rs (hunk 1 above)
Comment (xunilrj) [reply crates/rome_formatter/src/js/any/mod.rs]: Done.

File: crates/rslint_parser/Cargo.toml (hunk 2 above)
Comment (xunilrj) [reply crates/rslint_parser/Cargo.toml]: Done.

File: crates/rslint_parser/Cargo.toml (hunk 3 above)
Comment (xunilrj) [reply crates/rslint_parser/Cargo.toml]: Nothing... I removed it already.

File: crates/rslint_parser/src/parser.rs (hunk 4 above)
Comment (xunilrj) [reply crates/rslint_parser/src/parser.rs]: Actually, I don't wanna change this method. I will revert the changes.

File: crates/rslint_parser/src/syntax/expr.rs (hunk 6 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/expr.rs]: Done.

File: crates/rslint_parser/src/syntax/jsx/mod.rs (hunk 12 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs:78]: > Probably the error path is not covered in this PR

Yes. Still need to fix diagnostics with checkpoints. Then I will do all error checking.

File: xtask/codegen/js.ungram (hunk 13 above)
Comment (xunilrj) [reply xtask/codegen/js.ungram]: Done.

File: xtask/codegen/js.ungram (hunk 14 above)
Comment (xunilrj) [reply xtask/codegen/js.ungram]: Done.

File: xtask/codegen/js.ungram [hunk 15]
```diff
@@ -2136,3 +2139,52 @@ TsTypeArguments =
...
+
+JsxAnyElementName = 
+	JsxReferenceIdentifier
+	| JsxMemberExpression 
+
+JsxReferenceIdentifier = value: 'ident'
+
+// <a.test>	
+JsxMemberExpression = 
+	object: JsName
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply xtask/codegen/js.ungram:2186]: Mind if I push this to another PR?

[the remark this answers is not in the document]
Comment (ematipico) [reply xtask/codegen/js.ungram:2186]: Sure, no problem!

File: crates/rslint_parser/src/syntax/expr.rs (hunk 5 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/expr.rs]: I am checking now if JSX is enabled.

File: crates/rslint_parser/src/syntax/expr.rs (hunk 7 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/expr.rs]: Done

File: crates/rslint_parser/src/syntax/jsx/mod.rs (hunk 9 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs]: Good thinking. Done. Much better now.

**Comment by ematipico** [comment]
If possible, I would reword all the instances where we use "Jsx" in "JSX" (diagnostics, code, comments)

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        let checkpoint = p.checkpoint();
+        Self {
+            parser: p,
+            checkpoint,
+        }
+    }
+
+    pub fn unwrap(self) -> (&'a mut Parser<'b>, Checkpoint) {
+        (self.parser, self.checkpoint)
+    }
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs]: Done.

File: crates/rslint_parser/src/syntax/jsx/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs]: Done. But it is actually moving ```CheckpointedParser```.

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+// It is impossible to lookahead and guarantee that we are at a jsx expression,
+// so this function will checkpoint and rewind the parser on failures.
+pub(super) fn try_parse_jsx_expression(p: &mut Parser) -> ParsedSyntax {
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs]: Done. I like the consistency.

**Comment by xunilrj** [comment]
> If possible, I would reword all the instances where we use "Jsx" in "JSX" (diagnostics, code, comments)

Done. Where I thought the change made sense.

File: crates/rslint_parser/src/syntax/jsx/mod.rs (hunk 10 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs:87]: That is true.

File: crates/rslint_parser/src/syntax/jsx/mod.rs (hunk 8 above)
Comment (xunilrj) [reply crates/rslint_parser/src/syntax/jsx/mod.rs]: Mind if I push these and others changes to another PR? Maybe after JSX is more mature?

File: xtask/codegen/src/kinds_src.rs [hunk 18]
```diff
@@ -77,6 +77,8 @@ pub const KINDS_SRC: KindsSrc = KindsSrc {
         ("??=", "QUESTION2EQ"),
         ("@", "AT"),
         ("`", "BACKTICK"),
+        ("</", "L_ANGLE_SLASH"),
+        ("/>", "SLASH_R_ANGLE"),
```
Comment (ematipico) [inline xtask/codegen/src/kinds_src.rs:81]: Should we align the names? `L_ANGLE_SLASH` and `R_ANGLE_SLASH`

File: xtask/codegen/src/kinds_src.rs [hunk 19]
```diff
@@ -633,6 +645,8 @@ impl Field {
                     "??" => "nullish_coalescing",
                     "||" => "logical_or",
                     "&&" => "logical_and",
+                    "</" => "l_angle_slash",
+                    "/>" => "slash_r_angle",
```
Comment (ematipico) [inline xtask/codegen/src/kinds_src.rs:649]: Should we align the names? `l_angle_slash` and `r_angle_slash`

**Review by ematipico — APPROVED** [review]
Looks good to me. Left just two minor comments, no blockers. That's a good start!

File: crates/rslint_parser/src/syntax/jsx/mod.rs (hunk 10 above)
Comment (MichaReiser) [reply crates/rslint_parser/src/syntax/jsx/mod.rs:87]: We can figure this out when working on the children parsing

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+            ParsedSyntax::Present(self_closing_marker)
+        }
+        ParsedSyntax::Absent => ParsedSyntax::Absent,
+        _ => unreachable!("Unexpected present node returned"),
+    }
+}
+
+// <a ...> or <a ... />
+// ^          ^
+fn parse_jsx_element_head(p: &mut CheckpointedParser<'_, '_>, m: Marker) -> ParsedSyntax {
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:114]: Nit: I think you don't need to pass the `marker` in here. You can start it inside this function if needed. (For example, the method doesn't abandon the marker in case the parser isn't positioned at a `T![<]`

File: crates/rslint_parser/src/syntax/jsx/mod.rs [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+    if !p.parser.eat(T![<]) {
+        return ParsedSyntax::Absent;
+    }
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/jsx/mod.rs:117]: Isn't it possible to use `p.eat(T![<])` now that `CheckedParser` implements `Deref`

**Review by MichaReiser — APPROVED** [review]
